sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EquatableClassShouldBeSealed.cs

```diff
@@ -0,0 +1,116 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class EquatableClassShouldBeSealed : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4035";
+        private const string MessageFormat = "Make this class 'sealed' or implement 'IEqualityComparer<T>' instead.";
+        private const string EqualsMethodName = nameof(object.Equals);
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                    var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+                    if (classSymbol == null ||
+                        classSymbol.IsSealed ||
+                        classSymbol.IsStatic ||
+                        classDeclaration.Identifier.IsMissing)
+                    {
+                        return;
+                    }
+
+                    if (HasInvalidCombination(classSymbol))
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation()));
+                    }
+
+                }, SyntaxKind.ClassDeclaration);
+        }
+
+        private static bool IsValidEquatableInterfaceSymbol(INamedTypeSymbol namedTypeSymbol)
+        {
+            return namedTypeSymbol.ConstructedFrom.Is(KnownType.System_IEquatable_T) &&
+                namedTypeSymbol.TypeArguments.Length == 1;
+        }
+
+        private static bool IsValidEqualsMethodSymbol(IMethodSymbol methodSymbol)
```

Comment:
It's not clear what exactly "valid" means. Perhaps rename the method to make it more obvious?